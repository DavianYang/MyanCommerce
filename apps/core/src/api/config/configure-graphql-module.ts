import path from 'path';
import {
    GqlModuleOptions,
    GraphQLModule,
    GraphQLTypesLoader,
} from '@nestjs/graphql';
import { GraphQLSchema, buildSchema, printSchema } from 'graphql';
import { ConfigModule } from '../../config/config.module';
import { ServiceModule } from '../../service/service.module';
import { ConfigService } from '../../config/config.service';

export interface GraphQLApiOptions {
    apiType: 'shop' | 'admin';
    typePaths: string[];
    apiPath: string;
    debug: boolean;
    playground: boolean | any;
    resolverModule: Function;
}

/**
 * Dynamically generates a GraphQLModule according to the given config options.
 */
export function configureGraphQLModule(
    getOptions: (configService: ConfigService) => GraphQLApiOptions,
) {
    return GraphQLModule.forRootAsync({
        useFactory: (
            configService: ConfigService,
            typesLoader: GraphQLTypesLoader,
        ) => {
            return createGraphQLOptions(typesLoader, getOptions(configService));
        },
        inject: [ConfigService, GraphQLTypesLoader],
        imports: [ConfigModule, ServiceModule],
    });
}

async function createGraphQLOptions(
    typesLoader: GraphQLTypesLoader,
    options: GraphQLApiOptions,
): Promise<GqlModuleOptions> {
    const builtSchema = await buildSchemaForAPI();

    return {
        path: '/' + options.apiPath,
        typeDefs: printSchema(builtSchema),
        include: [options.resolverModule],
        playground: options.playground || false,
        debug: options.debug || false,
        context: (req: any) => req,
        cors: false,
    } as GqlModuleOptions;

    /**
     * Generates the server's GraphQL schema by combining:
     * 1. the default schema as defined in the source .graphql files specified by `typePaths`
     * 2. any custom fields defined in the config
     * 3. any schema extensions defined by plugins
     */
    async function buildSchemaForAPI(): Promise<GraphQLSchema> {
        // Paths must be normalized to use forward-slash separators.
        // See https://github.com/nestjs/graphql/issues/336
        const normalizedPaths = options.typePaths.map(p =>
            p.split(path.sep).join('/'),
        );
        const typeDefs = await typesLoader.mergeTypesByPaths(normalizedPaths);
        const schema = buildSchema(typeDefs);

        return schema;
    }
}
