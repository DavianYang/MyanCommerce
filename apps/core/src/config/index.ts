export * from './config.module';
export * from './config.service';
export * from './default.config';
export * from './merge-configs';
export * from './config.interface';
