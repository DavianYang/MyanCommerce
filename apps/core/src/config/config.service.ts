import { Injectable } from '@nestjs/common';
import { ConnectionOptions } from 'typeorm';
import { getConfig } from './config.helpers';
import { NSXLogger } from '@myancommerce/nsx-logger';
import {
    ApiOptions,
    MyanCommerceConfig,
    RuntimeMyanCommerceConfig,
} from './config.interface';

@Injectable()
export class ConfigService implements MyanCommerceConfig {
    private activeConfig: RuntimeMyanCommerceConfig;

    constructor() {
        this.activeConfig = getConfig();
    }

    get apiOptions(): Required<ApiOptions> {
        return this.activeConfig.apiOptions;
    }

    get logger(): NSXLogger {
        return this.activeConfig.logger;
    }

    get dbConnectionOptions(): ConnectionOptions {
        return this.activeConfig.dbConnectionOptions;
    }
}
