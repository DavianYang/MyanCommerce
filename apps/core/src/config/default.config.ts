import { RuntimeMyanCommerceConfig } from './config.interface';
import { DefaultLogger } from '../logger/default.logger';

/**
 * @description
 * The default configuration settings which are used if not explicitly overridden in the bootstrap() call.
 *
 * @docsCategory configuration
 */
export const defaultConfig: RuntimeMyanCommerceConfig = {
    logger: new DefaultLogger(),
    apiOptions: {
        hostname: '',
        port: 3000,
        adminApiPath: 'admin-api',
        adminApiPlayground: false,
        adminApiDebug: false,
        adminListQueryLimit: 1000,
        shopApiPath: 'shop-api',
        shopApiPlayground: false,
        shopApiDebug: false,
        shopListQueryLimit: 100,
        channelTokenKey: 'myancommerce-token',
        cors: {
            origin: true,
            credentials: true,
        },
        middleware: [],
    },
    dbConnectionOptions: {
        type: 'postgres',
    },
};
