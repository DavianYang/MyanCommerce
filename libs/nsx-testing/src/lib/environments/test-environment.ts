import * as dotenv from 'dotenv';
import { Request as HttpRequest, Response as HttpResponse } from 'express';
import { ConfigModuleOptions } from '@nestjs/config';

import { DefaultLogger } from '@myancommerce/nsx-logger';
import { LogLevel } from '@myancommerce/nsx-logger';
import {
    ApiOptions,
    AuthOptions,
    CometXConfig,
    GraphQLOptions,
    SocialAuthOptions,
} from '@myancommerce/core';
import { InMemoryCacheStrategy } from '@myancommerce/nsx-cache';
import {
    BcryptPasswordHashingStrategy,
    DefaultPasswordValidationStrategy,
    LocalAuthenticationStrategy,
} from '@myancommerce/nsx-auth';

dotenv.config({
    path: 'libs/nsx-testing/src/lib/.env.test',
});

const appConfig: ConfigModuleOptions = {
    isGlobal: true,
};

const apiConfig: ApiOptions = {
    hostname: process.env['DATABASE_HOST'],
    port: 3200,

    apiPath: 'api',
    apiPlayground: true, // turn this off for production
    apiDebug: true, // turn this off for production
};

const socialAuthConfig: SocialAuthOptions = {
    name: 'google',
    clientID: process.env['CLIENT_ID'] as string,
    clientSecret: process.env['CLIENT_SECRET'] as string,
};

const authConfig: AuthOptions = {
    jwtTokenSecret: process.env['JWT_SECRET'] as string,
    jwtTokenExpiry: process.env['JWT_EXPIRE_IN'] as string,
    jwtCookieExpiry: process.env['JWT_COOKIE_EXPIRES_IN'] as string,

    tokenMethod: 'cookie',
    cookieOptions: {
        secret: Math.random().toString(36).substring(3),
        httpOnly: true,
    },
    authTokenHeaderKey: 'cometx-auth-token',

    requireVerification: false,
    verificationTokenDuration: '7d',

    sessionCacheStrategy: new InMemoryCacheStrategy(),
    sessionDuration: '7d',
    sessionCacheTTL: 300,

    shopAuthenticationStrategy: [new LocalAuthenticationStrategy()],
    adminAuthenticationStrategy: [new LocalAuthenticationStrategy()],

    passwordHashingStrategy: new BcryptPasswordHashingStrategy(),
    passwordValidationStrategy: new DefaultPasswordValidationStrategy({
        minLength: 4,
    }),
};

const graphqlConfig: GraphQLOptions = {
    sortSchema: true,
    autoSchemaFile: 'apps/core/src/schema.gql',
    buildSchemaOptions: {
        numberScalarMode: 'integer',
    },
    cors: {
        credentials: true,
        origin: 'http://localhost:4200',
    },
    context: ({ req, res }: any) => ({
        request: req as HttpRequest,
        response: res as HttpResponse,
    }),
    formatError: (error: any) => ({
        code: error.extensions?.['code'] || 'SERVER_ERROR',
        name: error.extensions?.['exception']?.name || error.name,
        message:
            error.extensions?.['exception']?.response?.message || error.message,
        errLocations: apiConfig.apiDebug ? error.locations : undefined,
        errPath: apiConfig.apiDebug ? error.path : undefined,
        variables: apiConfig.apiDebug
            ? error.extensions?.['exception']?.variables
            : undefined,
        stacktrace: error.extensions?.['exception']?.stacktrace,
    }),
};
export const testEnvironment: CometXConfig = {
    appConfig,
    apiConfig,
    authConfig,
    socialAuthConfig,
    graphqlConfig,
    dbConnection: process.env['DATABASE_URL'] as string,
    logger: new DefaultLogger({
        level: LogLevel.Info,
        timestamp: true,
        defaultContext: 'CometX Server',
    }),
};
