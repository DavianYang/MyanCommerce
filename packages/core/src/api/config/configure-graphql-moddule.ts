import { GraphQLModule } from '@nestjs/graphql';
import { ConfigService } from 'config/config.service';

export interface GraphQLApiOptions {
    apiType: 'shop' | 'admin';
    typePaths: string[];
    debug: boolean;
    playground: boolean | any;
}

/**
 * Dynamically generates a GraphQLModule according to the given config options.
 */
export function configureGraphQLModule(
    getOptions: (configService: ConfigService) => GraphQLApiOptions,
) {
    return GraphQLModule.forRootAsync({
        useFactory: (configService: ConfigService) => {
            return {};
        },
    });
}
