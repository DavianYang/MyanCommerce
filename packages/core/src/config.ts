import path from 'path';
import * as dotenv from 'dotenv';
import { MyanCommerceConfig } from './config/myancommerce';
import { DefaultLogger } from './config/logger/default.logger';
import { LogLevel } from './config/logger/myancommerce.logger';

dotenv.config({ path: __dirname + `./.env.${process.env.NODE_ENV}` });

export const config: MyanCommerceConfig = {
    apiOptions: {
        port: 3000,
        adminApiPath: 'admin-api',
        adminApiPlayground: true, // turn this off for production
        adminApiDebug: true, // turn this off for production
        shopApiPath: 'shop-api',
        shopApiPlayground: {
            settings: {
                'request.credentials': 'include',
            } as any,
        }, // turn this off for production
        shopApiDebug: true, // turn this off for production
    },
    dbConnectionOptions: {
        type: 'postgres',
        synchronize: false, // turn off for production
        logging: false,
        database: process.env.DATABASE_NAME,
        host: process.env.DATABASE_HOST,
        port: 5433,
        username: process.env.DATABASE_USER,
        password: process.env.DATABASE_PASSWORD,
        migrations: [path.join(__dirname, '../migrations/*.ts')],
    },
    logger: new DefaultLogger({ level: LogLevel.Info, timestamp: false }),
};
