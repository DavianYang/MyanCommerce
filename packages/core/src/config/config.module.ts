import { Module, OnApplicationBootstrap, OnApplicationShutdown } from '@nestjs/common';
import { ModuleRef } from '@nestjs/core';
import { ConfigService } from './config.service';

@Module({
    providers: [ConfigService],
    exports: [ConfigService]
})
export class ConfigModule implements OnApplicationBootstrap, OnApplicationShutdown {
    constructor(private configService: ConfigService, private moduleref: ModuleRef){}

    async onApplicationBootstrap() {
        
    }

    async onApplicationShutdown(signal?: string) {
        
    }

}