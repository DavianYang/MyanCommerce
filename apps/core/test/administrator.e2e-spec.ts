import {
    GetAdministrators,
    CreateAdministrator,
    Administrator,
    GetAdministrator,
    UpdateAdministrator,
    DeleteAdministrator,
    DeletionResult,
} from '@myancommerce/generated';
import { createTestEnvironment } from '@myancommerce/testing';
import gql from 'graphql-tag';
import { testConfig } from './config/test-environment';

describe('Adminstrator resolver', () => {
    const { server, adminClient } = createTestEnvironment(testConfig());
    let createdAdmin: Administrator.Fragment;

    beforeAll(async () => {
        await server.init({ adminCount: 1 });
    });

    afterAll(async () => {
        await server.destroy();
    });

    it('get adminstrators', async () => {
        const result = await adminClient.query<
            GetAdministrators.Query,
            GetAdministrators.Variables
        >(GET_ADMINISTRATORS);

        expect(result.administrators.items.length).toBe(1);
        expect(result.administrators.totalItems).toBe(1);
    });

    it('create administrator', async () => {
        const result = await adminClient.query<
            CreateAdministrator.Mutation,
            CreateAdministrator.Variables
        >(CREATE_ADMINISTRATOR, {
            input: {
                emailAddress: 'hello@test.com',
                firstName: 'Mg',
                lastName: 'Mg',
                roleIds: ['1'],
            },
        });

        createdAdmin = result.createAdministrator;
        expect(createdAdmin).toMatchSnapshot();
    });

    it('get administrator', async () => {
        const result = await adminClient.query<
            GetAdministrator.Query,
            GetAdministrator.Variables
        >(GET_ADMINISTRATOR, { id: createdAdmin.id });

        expect(result.administrator).toEqual(createdAdmin);
    });

    it('update administrator', async () => {
        const result = await adminClient.query<
            UpdateAdministrator.Mutation,
            UpdateAdministrator.Variables
        >(UPDATE_ADMINISTRATOR, {
            id: createdAdmin.id,
            input: {
                emailAddress: 'testemail@gmail.com',
                firstName: 'New First',
                lastName: 'New Last',
                roleIds: ['2'],
            },
        });

        expect(result.updateAdministrator).toMatchSnapshot();
    });

    it('update administrator with partial input', async () => {
        const result = await adminClient.query<
            UpdateAdministrator.Mutation,
            UpdateAdministrator.Variables
        >(UPDATE_ADMINISTRATOR, {
            id: createdAdmin.id,
            input: {
                emailAddress: 'newtestemail@gmail.com',
            },
        });

        expect(result.updateAdministrator.emailAddress).toBe(
            'newtestemail@gmail.com',
        );
        expect(result.updateAdministrator.firstName).toBe('New First');
        expect(result.updateAdministrator.lastName).toBe('New Last');
    });

    it('delete administrator', async () => {
        const { administrators: before } = await adminClient.query<
            GetAdministrators.Query,
            GetAdministrators.Variables
        >(GET_ADMINISTRATORS);

        expect(before.totalItems).toBe(2);

        const { deleteAdministrator } = await adminClient.query<
            DeleteAdministrator.Mutation,
            DeleteAdministrator.Variables
        >(DELETE_ADMINISTRATOR, { id: createdAdmin.id });

        expect(deleteAdministrator.result).toBe(DeletionResult.Deleted);

        const { administrators: after } = await adminClient.query<
            GetAdministrators.Query,
            GetAdministrators.Variables
        >(GET_ADMINISTRATORS);

        expect(after.totalItems).toBe(1);
    });
});

export const ADMINISTRATOR_FRAGMENT = gql`
    fragment Administrator on Administrator {
        id
        firstName
        lastName
        emailAddress
    }
`;

export const GET_ADMINISTRATORS = gql`
    query GetAdministrators {
        administrators {
            items {
                ...Administrator
            }
            totalItems
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const GET_ADMINISTRATOR = gql`
    query GetAdministrator($id: ID!) {
        administrator(id: $id) {
            ...Administrator
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const CREATE_ADMINISTRATOR = gql`
    mutation CreateAdministrator($input: CreateAdministratorInput!) {
        createAdministrator(input: $input) {
            ...Administrator
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const UPDATE_ADMINISTRATOR = gql`
    mutation UpdateAdministrator($id: ID!, $input: UpdateAdministratorInput!) {
        updateAdministrator(id: $id, input: $input) {
            ...Administrator
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const DELETE_ADMINISTRATOR = gql`
    mutation DeleteAdministrator($id: ID!) {
        deleteAdministrator(id: $id) {
            message
            result
        }
    }
`;
