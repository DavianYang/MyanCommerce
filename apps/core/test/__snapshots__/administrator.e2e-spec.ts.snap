// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adminstrator resolver create administrator 1`] = `
Object {
  "emailAddress": "hello@test.com",
  "firstName": "Mg",
  "id": "2",
  "lastName": "Mg",
}
`;

exports[`Adminstrator resolver update administrator 1`] = `
Object {
  "emailAddress": "testemail@gmail.com",
  "firstName": "New First",
  "id": "2",
  "lastName": "New Last",
}
`;
